packages/timeline-state-resolver/src/integrations/singularLive/index.ts

```diff
@@ -186,138 +122,121 @@ export class SingularLiveDevice extends DeviceWithState<SingularLiveState, Devic
-	): Array<Command> {
-		const commands: Array<Command> = []
+	diffStates(
+		oldSingularLiveState: SingularLiveState | undefined,
+		newSingularLiveState: SingularLiveState,
+		_mappings: Mappings<unknown>,
+		_time: number
+	): SingularLiveCommandContext[] {
+		const commands: Array<SingularLiveCommandContext> = []
 
 		_.each(newSingularLiveState.compositions, (composition: SingularComposition, compositionName: string) => {
-			const oldComposition = oldSingularLiveState.compositions[compositionName]
+			const oldComposition = oldSingularLiveState?.compositions?.[compositionName]
 			if (!oldComposition) {
 				// added!
 				commands.push({
 					timelineObjId: composition.timelineObjId,
-					commandName: 'added',
-					content: literal<SingularLiveControlNodeCommandContent>({
-						subCompositionName: compositionName,
-						state: composition.controlNode.state,
-						payload: composition.controlNode.payload,
-					}),
 					context: `added: ${composition.timelineObjId}`,
-					layer: compositionName,
+					command: {
+						commandName: 'added',
+						content: literal<SingularLiveControlNodeCommandContent>({
+							subCompositionName: compositionName,
+							state: composition.controlNode.state,
+							payload: composition.controlNode.payload,
+						}),
+						layer: compositionName,
+					},
 				})
 			} else {
 				// changed?
 				if (!_.isEqual(oldComposition.controlNode, composition.controlNode)) {
 					// changed!
 					commands.push({
 						timelineObjId: composition.timelineObjId,
-						commandName: 'changed',
-						content: literal<SingularLiveControlNodeCommandContent>({
-							subCompositionName: compositionName,
-							state: composition.controlNode.state,
-							payload: composition.controlNode.payload,
-						}),
 						context: `changed: ${composition.timelineObjId}  (previously: ${oldComposition.timelineObjId})`,
-						layer: compositionName,
+						command: {
+							commandName: 'changed',
+							content: literal<SingularLiveControlNodeCommandContent>({
+								subCompositionName: compositionName,
+								state: composition.controlNode.state,
+								payload: composition.controlNode.payload,
+							}),
+							layer: compositionName,
+						},
 					})
 				}
 			}
 		})
 		// removed
-		_.each(oldSingularLiveState.compositions, (composition: SingularComposition, compositionName) => {
-			const newComposition = newSingularLiveState.compositions[compositionName]
-			if (!newComposition) {
-				// removed!
-				commands.push({
-					timelineObjId: composition.timelineObjId,
-					commandName: 'removed',
-					content: literal<SingularLiveControlNodeCommandContent>({
-						subCompositionName: compositionName,
-						state: 'Out',
-					}),
-					context: `removed: ${composition.timelineObjId}`,
-					layer: compositionName,
-				})
-			}
-		})
+		if (oldSingularLiveState) {
+			_.each(oldSingularLiveState.compositions, (composition: SingularComposition, compositionName) => {
```

Comment:
minor note: Perhaps we should refactor away the underscore dependencies as well?

Reply:
I almost did, but there is a `_.isEqual` somewhere here which I'm not aware of a direct replacement for, so didn't feel it worth changing these other references